Reconnect websockets when they close unexpectedly

djjuhasz opened the pull request:
Fixes #1057.

Attempt to reconnect the ingest and storage monitor websockets when the connection is unexpected closed at the server end (e.g. on server restart). My tests so far indicate that this change fixes the problem I've been having with the Dashboard hanging intermittently and throwing Vite "write EPIPE" errors.

djjuhasz committed:
Reconnect websockets when they close unexpectedly

Fixes #1057.

Attempt to reconnect the ingest and storage monitor websockets when the
connection is unexpected closed at the server end (e.g. on server
restart). My tests so far indicate that this change fixes the problem
I've been having with the Dashboard hanging intermittently and throwing
Vite "write EPIPE" errors.

The reconnection logic does exponential backoff and jitter to avoid
overwhelming a slow server with too many reconnection requests, and
limits the total number of reconnection attempts.

The WebSocket connection state is managed using Pinia to persist the
connections between page updates.

jraddaoui commented on dashboard/src/client.ts:
```diff
@@ -48,6 +48,18 @@ function connectIngestMonitor() {
+    console.error("Ingest monitor socket closed");
+    setTimeout(() => {
+      console.log("Reconnecting ingest monitor socket...");
+      client.ingest.ingestMonitorRequest().then(() => {
```
I'd consider putting this initial request inside the `connectIngestMonitor` function directly, and the same for storage. It would reduce some duplication, leaving the calls in App.vue just with `client.connectIngestMonitor();` and `client.connectStorageMonitor();` inside the if, and similar here on close.

https://github.com/artefactual-sdps/enduro/blob/main/dashboard/src/App.vue#L19-L24

djjuhasz replied:
@jraddaoui I don't understand what you mean.  The `connectIngestMonitor()` method is autogenerated, I'm not sure how we would modify the function put the `ingestMonitorRequest()` call "inside" it. :confused:

djjuhasz replied:
Oh, I see. `connectIngestMonitor()` isn't auto-generated, `ingestMonitorRequest()` is auto-generated.

jraddaoui reviewed (approved):
LGTM @djjuhasz, thanks for looking into this! Just a minor suggestion to reduce duplication.

djjuhasz commented:
@jraddaoui I totally rewrote the monitor reconnection logic to include exponential backoff with jitter, and an attempt limit.  I've also moved the ingest and monitor state to Pinia to preserve the WebSocket connections through page updates.  I've also added tests for all of the reconnection logic.

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+  close(): void;
+}
+
+export class MonitorConnection {
```
I'd make this an abstract class with an abstract dial method and implement `MonitorConnector` directly. That will force all subclasses to implement `MonitorConnector` even if it's forgotten and it will reduce some minor duplication (see below).

```ts
export abstract class MonitorConnection implements MonitorConnector {
  // Existing code.
  
  abstract dial(): Promise<void>;
}
```

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+    this.socket.onerror = () => {
+      console.error(`${this.type} monitor socket error`);
+    };
+  }
```
With the abstract dial method you could move the `onclose` set up here too:

```ts
connectSocket(url: string): void {
  this.socket = new WebSocket(url);

  this.socket.onopen = () => {
    console.log(`${this.type} monitor socket connected`);
  };
  this.socket.onerror = () => {
    console.error(`${this.type} monitor socket error`);
  };

  // Reconnect on close.
  this.socket.onclose = () => {
    console.error(`${this.type} monitor socket closed, reconnecting...`);
    this.dial();
  };
}
```

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+export class IngestMonitorConnection
+  extends MonitorConnection
+  implements MonitorConnector
+{
```
This will only need to extend MonitorConnection:

```ts
export class IngestMonitorConnection extends MonitorConnection {}
```

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+              this.socket.onclose = () => {
+                console.error("ingest monitor socket closed, reconnecting...");
+                this.dial();
+              };
```
And this could be removed.

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+export class StorageMonitorConnection
+  extends MonitorConnection
+  implements MonitorConnector
+{
```
Same.

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+              this.socket.onclose = () => {
+                console.error("storage monitor socket closed, reconnecting...");
+                this.dial();
+              };
```
Same.

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+  isConnected(): boolean;
+  getWebSocketURL(url: string): string;
+  close(): void;
+}
```
Actually, with the abstract class I think we could remove this interface entirely, the `MonitorConnection` class will define the contract.

jraddaoui commented on dashboard/src/monitor.ts:
```diff
@@ -0,0 +1,180 @@
+    }
+  }
+
+  async connectSocket(url: string): Promise<void> {
```
Why is this method async and returning a promise?

```ts
connectSocket(url: string): void {}
```

That would work, right? Then we'll need to update the dial calls:

```ts
this.connectSocket(this.url);
if (this.socket) {}
```

djjuhasz replied:
I assumed `new Websocket()` would connect asynchronously and return a promise, but it looks like creating the socket is blocking (and pretty much instantaneous), then the `onopen` handler fires asynchronously when the connection is established. :thinking:

When looking this up I found a great example of a web socket client class: https://websocket.org/guides/languages/javascript/.  I wish I'd found that before trying to write this whole class myself. :facepalm:

jraddaoui reviewed (changes requested):
Thanks @djjuhasz, looks great to me, nice work!

We could simplify a little the monitor classes by using an abstract class, but it's not a huge deal.

djjuhasz commented:
@jraddaoui I think I've addressed all your feedback and this is ready for another look.  I also stopped exporting the "MonitorConnect" class, and changed the `isConnected()` method to an `isConnected` property, following the example of: https://websocket.org/guides/languages/javascript/

jraddaoui reviewed (approved):
Thanks @djjuhasz!
